### PR TITLE
Offline cask search and listing all available casks 

### LINKED ISSUE #3080: Brew cask search shouldn't fail when offline
- [x] Confirmed this is a problem with running a `brew` command and not `brew install`ing or the post-install behaviour of one or more formulae? If it's a formulae-specific problem please file this issue at https://github.com/Homebrew/homebrew-core/issues/new
- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed all issues and retried your prior step?
- [x] Ran `brew config` and `brew doctor` and included their output with your issue?

<details><summary>Output of `brew doctor` and `brew config`</summary>

```
> brew doctor
Your system is ready to brew.

> brew config
HOMEBREW_VERSION: 1.3.1-77-g97c75bfd
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 97c75bfd40084cfa2512fe55a7b292886ff6d023
Last commit: 3 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 280a946338f565223ba6480e84f3cdf2ea7b9fa2
Core tap last commit: 14 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: quad-core 64-bit haswell
Homebrew Ruby: 2.0.0-p648
Clang: 8.0 build 800
Git: 2.2.1 => /usr/local/git/bin/git
Perl: /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
Java: 1.8.0_66
macOS: 10.11.6-x86_64
Xcode: N/A
CLT: 8.2.0.0.1.1480973914
X11: N/A
```

</details>

### What you were trying to do (and why)

Get the list of available casks by running `brew cask search`  (without search query)

### What happened (include command output)

It failed, because I was offline.

<details><summary>Error output</summary>

```
Error: curl failed!
curl: (6) Could not resolve host: api.github.com
curl: (3) <url> malformed

Follow the instructions here:
  https://github.com/caskroom/homebrew-cask#reporting-bugs
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:226:in `raise_api_error'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:184:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:284:in `search'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:239:in `search_code'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:18:in `search_remote'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:44:in `search'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:5:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/abstract_command.rb:35:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:97:in `run_command'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:167:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:131:in `run'
/usr/local/Homebrew/Library/Homebrew/cmd/cask.rb:8:in `cask'
/usr/local/Homebrew/Library/Homebrew/brew.rb:95:in `<main>'
```

</details>

### What you expected to happen

I would expect `brew cask search` to work similarly to the `brew search` command:

```
brew search, -S:
    Display all locally available formulae for brewing (including tapped ones).
    No online search is performed if called without arguments.
```

So I expected `brew cask search` to return a list of cask tokens **without performing online search**. But even if it does, it shouldn't completely fail, but rather return an empty result in "Remote matches".

### Step-by-step reproduction instructions (by running `brew` commands)

1. turn off internet connection
2. run `brew cask search`

----

The failing line is [here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cask/lib/hbc/cli/search.rb#L44), would be nice if it just returned an empty list of results when it cannot search online. It could output the error message to stderr.

Another thing is that I think it shouldn't search online at all, when run without arguments. I found [this piece of code](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cask/lib/hbc/scopes.rb#L16-L22) that seems to do what I expect:

```
Tap.flat_map { |t| t.cask_files.map { |p| "#{t.name}/#{File.basename(p, ".rb")}" } }
```

(by the way, how can I call `Hbc.all_tokens` in `brew irb` directly?)

### PR DESCRIPTION
**Pull request by laughedelic**
Fixes #3080 

---- 

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
  _See below_
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Here are two changes for fixing #3080:
* a5640fd handles Github request failure, so that the whole search doesn't fail and returns empty "remote matches" instead
* 00803b1 makes `brew cask search` to output the list of all cask tokens without performing any search (analogously to `brew search`)

I tried to improve tests, but I have some problems with it and decided to submit a WIP pull-request and ask for help here:

1. I noticed that there was already a test for the expected behaviour described in #3080: ["lists all available Casks with no search term"](https://github.com/Homebrew/brew/blob/a5640fdfeeea7347e51c14abac481f4b210527ab/Library/Homebrew/test/cask/cli/search_spec.rb#L31-L35). But it fails **only** when I run tests with `--online` _and_ turned off network connection. I don't really understand why it passes without `--online` and no network. Are any network requests silently ignored without `--online` flag?
2. I noticed usage of `:needs_network` in other tests, but I don't know how to use it here: there should be tests both with and without network
3. Is there a way to specify that a test has to run successfully _without_ network? `:doesnt_need_network`? or a way to mock such situation?

### CODE REVIEW

File: Library/Homebrew/cask/lib/hbc/cli/search.rb [hunk 1]
```diff
@@ -2,8 +2,12 @@ module Hbc
   class CLI
     class Search < AbstractCommand
       def run
-        results = self.class.search(*args)
-        self.class.render_results(*results)
+        if args.empty?
+          puts Formatter.columns(CLI.nice_listing(Hbc.all_tokens))
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: This should only be called if there is no connection and should probably be handled in `::search`.

Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: This should only be called if there is no connection and should probably be handled in `::search`.

File: Library/Homebrew/cask/lib/hbc/cli/search.rb [hunk 2]
```diff
@@ -15,8 +19,18 @@ def self.extract_regexp(string)
...
       def self.search_remote(query)
-        matches = GitHub.search_code(user: "caskroom", path: "Casks",
-                                     filename: query, extension: "rb")
+        matches = begin GitHub.search_code(
+          user: "caskroom",
+          path: "Casks",
+          filename: query,
+          extension: "rb"
+        )
+        rescue Exception => e
```
Comment (laughedelic) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: I changed it to `StandardError`. The error raised is [this](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/github.rb#L220-L226), can it be made more specific?

File: Library/Homebrew/cask/lib/hbc/cli/search.rb [hunk 3]
```diff
@@ -15,8 +19,18 @@ def self.extract_regexp(string)
       end
 
       def self.search_remote(query)
-        matches = GitHub.search_code(user: "caskroom", path: "Casks",
-                                     filename: query, extension: "rb")
+        matches = begin GitHub.search_code(
```
Comment (laughedelic) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: 15d2a9c like this?

File: Library/Homebrew/cask/lib/hbc/cli/search.rb (hunk 1 above)
Comment (laughedelic) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: The whole point of this change is that `brew cask search` (without arguments) _should not_ perform online search, same way is `brew search` does (see [here](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/search.rb#L26-L27)). 
You can still do `brew cask search ""` which will do a normal search returning all casks.

File: Library/Homebrew/cask/lib/hbc/cli/search.rb (hunk 2 above)
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: Ok, use `Github::Error` then.

Comment (laughedelic) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: Done in 2cc6ef4

**Comment by reitermarkus** [comment]
Ok, now for testing his, you want `allow(GitHub).to receive(:search_code).and raise_error(GitHub::Error.new("curl failed! reason“))`, followed by the actual test.

**Comment by laughedelic** [comment]
@reitermarkus hi! I added 2 new tests:
* for when online search failed: check the expected output and expected error message
* for empty matches output: it should be empty when stdout is not TTY

And modified the test for casks listing when `brew cask search` is called with no search term to ensure it's not affected by an online search failure.

I also unified formatting for some other tests in 68c837f, I hope you don't mind.

File: Library/Homebrew/cask/lib/hbc/cli/search.rb [hunk 4]
```diff
@@ -15,8 +19,18 @@ def self.extract_regexp(string)
...
+        matches = begin
+          GitHub.search_code(
+            user: "caskroom",
+            path: "Casks",
+            filename: query,
+            extension: "rb",
+          )
+        rescue GitHub::Error => e
+          onoe e
+          $stderr.puts e.backtrace
```
Comment (laughedelic) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: One doubt here: probably outputting the backtrace is a bit too much? `onoe e` will output in this case 

```
Error: curl failed!
curl: (6) Could not resolve host: api.github.com
curl: (3) <url> malformed
```

And the backtrace is not helpful for the user:

```
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:226:in `raise_api_error'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:184:in `open'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:284:in `search'
/usr/local/Homebrew/Library/Homebrew/utils/github.rb:239:in `search_code'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:23:in `search_remote'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:58:in `search'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/search.rb:8:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli/abstract_command.rb:35:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:97:in `run_command'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:167:in `run'
/usr/local/Homebrew/Library/Homebrew/cask/lib/hbc/cli.rb:131:in `run'
/usr/local/Homebrew/Library/Homebrew/cmd/cask.rb:8:in `cask'
/usr/local/Homebrew/Library/Homebrew/brew.rb:95:in `<main>'
```

----

Probably the error message could be improved too (in `Library/Homebrew/utils/github.rb`). Like matching that error output from `curl`, checking the network connection and then doing `onoe 'Check your network connection'`. What do you think?

Comment (laughedelic) [reply Library/Homebrew/cask/lib/hbc/cli/search.rb]: I removed the stacktrace in 2a3f83b8d4ceeca370a60bbcb494990a2bd20801 and changed error to warning (as it doesn't fail the whole search). Now it looks like this:

```
$ brew cask search "foo-bar"
Warning: Online search failed: curl failed!
curl: (6) Could not resolve host: api.github.com
curl: (3) <url> malformed

No Cask found for "foo-bar".
```

Online/Github/Remote search?

**Comment by laughedelic** [comment]
@reitermarkus could you review this again?

File: Library/Homebrew/cask/lib/hbc/cli/search.rb [hunk 5]
```diff
@@ -15,8 +19,17 @@ def self.extract_regexp(string)
...
-                                     filename: query, extension: "rb")
+        matches = begin
+          GitHub.search_code(
+            user: "caskroom",
+            path: "Casks",
+            filename: query,
+            extension: "rb",
+          )
+        rescue GitHub::Error => error
+          opoo "Online search failed: #{error}\n"
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/cli/search.rb]: I'd rephrase this to “Error searching on GitHub:”.

File: Library/Homebrew/test/cask/cli/search_spec.rb [hunk 6]
```diff
@@ -22,16 +22,38 @@
...
-    }.to output("No Cask found for \"foo-bar-baz\".\n").to_stdout.as_tty
+    }.to output(<<-EOS.undent).to_stdout.as_tty
+      No Cask found for "foo-bar-baz".
+    EOS
+  end
+
+  it "doesn't output anything to non-TTY stdout when there are no matches" do
+    expect {
+      Hbc::CLI::Search.run("foo-bar-baz")
+    }.to not_to_output.to_stdout
```
Comment (reitermarkus) [inline Library/Homebrew/test/cask/cli/search_spec.rb]: `.not_to output.to_stdout`

Comment (laughedelic) [reply Library/Homebrew/test/cask/cli/search_spec.rb]: I tried to change it to `.not_to output.to_stdout.or output.to_stderr`, but then got the following error:

> `expect(...).not_to matcher.or matcher` is not supported, since it creates a bit of an ambiguity. Instead, define negated versions of whatever matchers you wish to negate with `RSpec::Matchers.define_negated_matcher` and use `expect(...).to matcher.or matcher`.

There are two matchers there: no output to stdout and no output to stderr. How should I rewrite it?

File: Library/Homebrew/cask/lib/hbc/cli/search.rb (hunk 5 above)
Comment (laughedelic) [reply Library/Homebrew/cask/lib/hbc/cli/search.rb]: Done in 68dd0ac

File: Library/Homebrew/test/cask/cli/search_spec.rb (hunk 6 above)
Comment (reitermarkus) [reply Library/Homebrew/test/cask/cli/search_spec.rb]: Ok, then make it `.to not_output.to_stderr`, and add another negated matcher like `not_to_output`.

File: Library/Homebrew/test/cask/cli/search_spec.rb [hunk 7]
```diff
@@ -22,16 +22,38 @@
...
+    .and not_to_output.to_stderr
   end
 
-  it "lists all available Casks with no search term" do
+  it "lists all Casks available offline with no search term" do
+    allow(GitHub).to receive(:search_code).and_raise(GitHub::Error.new("reason"))
     expect {
       Hbc::CLI::Search.run
     }.to output(/local-caffeine/).to_stdout.as_tty
+    .and not_to_output.to_stderr
```
Comment (reitermarkus) [inline Library/Homebrew/test/cask/cli/search_spec.rb]: Make the expect a single line and put the `.to` and `.and` line below it.

Comment (laughedelic) [reply Library/Homebrew/test/cask/cli/search_spec.rb]: Done in 8e5ad9a. Should I change the rest of the tests to this formatting too?

File: Library/Homebrew/test/cask/cli/search_spec.rb (hunk 6 above)
Comment (laughedelic) [reply Library/Homebrew/test/cask/cli/search_spec.rb]: Sorry, I'm not sure what you mean. It seems that it is already as you say, two negated conditions: `¬A & ¬B`:

```ruby
.to not_to_output.to_stdout
.and not_to_output.to_stderr
```

I changed the formatting in 8e5ad9a though.

File: Library/Homebrew/cask/lib/hbc/cli/search.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/cask/lib/hbc/cli/search.rb]: The last thing here would be to update the man page to mention this, other than that it looks good to me. (Make sure to run `brew man` afterwards.)

**Comment by laughedelic** [comment]
@reitermarkus done in 50db7db 👌

**Comment by reitermarkus** [comment]
Great work here, @laughedelic!

**Comment by laughedelic** [comment]
Thanks!
